debuginfo/src/elf.rs

```diff
@@ -0,0 +1,185 @@
+    for phdr in &elf.program_headers {
+        if phdr.p_type != elf::program_header::PT_NOTE {
+            continue;
+        }
+
+        let offset = phdr.p_offset as usize;
+        let mut notes = elf::note::NoteIterator {
+            data,
+            offset,
+            size: offset + phdr.p_filesz as usize,
+            ctx: (phdr.p_align as usize, elf.ctx),
+        };
+
+        if let Some(note) = find_elf_note(notes, elf::note::NT_GNU_BUILD_ID) {
+            return Some(note);
+        }
+    }
+
+    // Next, load the SHT_NOTE section called ".note.gnu.build-id". It contains
+    // the same data as the PT_NOTE program header. Again, swallow all errors
+    // and fall through if reading the section is not possible.
+    let sh_type = elf::section_header::SHT_NOTE;
+    if let Some(section) = find_elf_section(elf, data, sh_type, ".note.gnu.build-id") {
+        let offset = section.header.sh_offset as usize;
+        let mut notes = elf::note::NoteIterator {
+            data,
+            offset,
+            size: offset + section.header.sh_size as usize,
+            ctx: (section.header.sh_addralign as usize, elf.ctx),
+        };
+
+        if let Some(note) = find_elf_note(notes, elf::note::NT_GNU_BUILD_ID) {
+            return Some(note);
+        }
+    }
+
+    None
+}
+
+/// Converts an ELF object identifier into a `Uuid`
+///
+/// The identifier data is first truncated or extended to match 16 byte size of
+/// Uuids. If the data is declared in little endian, the first three Uuid fields
+/// are flipped to match the big endian expected by the breakpad processor.
+fn create_elf_uuid(identifier: &[u8], little_endian: bool) -> Option<Uuid> {
+    // Make sure that we have exactly UUID_SIZE bytes available
+    let mut data = [0 as u8; UUID_SIZE];
+    let len = cmp::min(identifier.len(), UUID_SIZE);
+    data[0..len].copy_from_slice(&identifier[0..len]);
+
+    if little_endian {
+        // The file ELF file targets a little endian architecture. Convert to
+        // network byte order (big endian) to match the Breakpad processor's
+        // expectations. For big endian object files, this is not needed.
+        data[0..4].reverse(); // uuid field 1
+        data[4..6].reverse(); // uuid field 2
+        data[6..8].reverse(); // uuid field 3
+    }
+
+    Uuid::from_bytes(&data).ok()
+}
+
+/// Tries to obtain the object UUID of an ELF object.
+///
+/// As opposed to Mach-O, ELF does not specify a unique ID for object files in
+/// its header. Compilers and linkers usually add either `SHT_NOTE` sections or
+/// `PT_NOTE` program header elements for this purpose.
+///
+/// If neither of the above are present, this function will hash the first page
+/// of the `.text` section (program code) to synthesize a unique ID. This is
+/// likely not a valid UUID since was generated off a hash value.
+///
+/// If all of the above fails, the UUID will be `None`.
+pub fn get_elf_uuid(elf: &elf::Elf, data: &[u8]) -> Option<Uuid> {
```

Comment:
This is the entrypoint to calculating an ELF uuid. See `find_build_id` for the logic of resolving the original build ID and `create_elf_uuid` for the endianness normalization.